manual/tools/caml_tex2.ml

```
@@ -34,6 +34,127 @@ let (~!) =
+    | Ok -> Printf.fprintf ppf "an ok"
+    | Warning n -> Printf.fprintf ppf "a warning %d" n
+
+  (** {2 Exceptions } *)
+  exception Parsing_error of kind * string
+  type unexpected_report = {source:string; expected:status; got:status}
+  exception Unexpected_status of unexpected_report
+
+  let print_unexpected {source; expected; got} =
+    if expected = Ok then
+      Printf.eprintf
+        "Error when evaluating a caml_example environment in %s:\n\
+         unexpected %a.\n\
+         If %a status was expected, add an [@@expect %a] annotation.\n"
+        source
+        pp_status got
+        pp_a_status got
+        pp_status got
+    else
+      Printf.eprintf
+        "Error when evaluating a guarded caml_example environment in %s:\n \
+         %a status was expected, got %a status.\n\
+         If %a states was in fact expected, change the status annotation to\
+         [@@expect %a].\n"
+        source
+        pp_a_status expected
+        pp_a_status got
+        pp_a_status got
+        pp_status got;
+    flush stderr
+
+  let print_parsing_error k s =
+    match k with
+    | Option ->
+        Printf.eprintf
+          "Unknown caml_example option: [%s].\n\
+           Supported options are \"ok\",\"error\", or \"warning=n\" (with n \
+           a warning number).\n" s
+    | Annotation ->
+        Printf.eprintf
+          "Unknown caml_example phrase annotation: [@@expect %s].\n\
+           Supported annotations are [@@expect ok], [@@expect error],\n\
+           and [@@expect warning n] (with n a warning number).\n" s
+
+
+  (** {2 Output analysis} *)
+  let catch_error s =
+    if string_match ~!{|Error:|} s 0 then Some Error else None
+
+  let catch_warning s =
+    if string_match ~!{|Warning \([0-9]+\):|} s 0 then
+      Some (Warning (int_of_string @@ matched_group 1 s))
+    else
+      None
+
+  let status s = match catch_warning s, catch_error s with
+    | Some w, _ -> w
+    | _, Some e -> e
```

In the current implementation, the two clauses should be commutative since simultaneous multiple warnings and error are not handled.  

Should they be handled? it would increase the complexity of the parsing for options and I am not sure how useful it would be. Maybe detecting the situation and raising an error would be worthwile?

Anyway I am perfectly fine with adding a `None` to choose an arbitrary ordering of these clauses.